packages/material-ui/src/useMediaQuery/unstable_useMediaQuery.js

```diff
@@ -0,0 +1,60 @@
+/* eslint-disable camelcase */
+
+import React from 'react';
+
+// This variable will be true once the server-side reconciliation is done.
+let didMount = false;
```

Comment:
Faking instance variables can usually be achieved with `useRef`. Why is a singleton used here? Standard `useIsMounted`:
```ts
function useIsMounted(): React.RefObject<boolean> {
  const mountedRef = React.useRef(false);

  React.useEffect(() => {
    mountedRef.current = true;

    return () => {
      mountedRef.current = false;
    }
  }, []);

  return mountedRef;
}
```

Reply:
It's a performance optimisation. I would rename it: `didMountOnce`. To gain more insight about what it does, you can remove the logic and see the corresponding test case fail. It will render more times that it need.

Reply:
What performance issue did you encounter and by how much did you solve it? 

Do you have any additional resources that explain that reconciliation is done in `useEffect`? As far as I know reconciliation is a repeated process in react so `firstReconciliationDone` is probably the better name.

I'm not familiar with SSR but do they really throw away the hole tree per request? What if the server keeps the component tree mounted and just runs reconciliation for every request. This could not be supported with this implementation.

This is now only usable once per application because of the singleton. Any additional usage will potentially consider itself mounted when it's not.

I did replace the logic with my `useIsMounted` the only thing that failed is "should try to reconciliate only the first time". I'm honestly confused by this requirement. I have never seen the singleton components in react which is the only use case for this hook.

Reply:
`firstReconciliation` would be better 👍. The goal is to avoid the component to render twice. Without this logic, it will render once with the default value and a second time with the right value. This is what happens with the current withWidth implementation, it's wasteful: 
> You may call setState() immediately in componentDidMount(). It will trigger an extra rendering, but it will happen before the browser updates the screen. This guarantees that even though the render() will be called twice in this case, the user won’t see the intermediate state. Use this pattern with caution because it often causes performance issues. **In most cases, you should be able to assign the initial state in the constructor() instead**. It can, however, be necessary for cases like modals and tooltips when you need to measure a DOM node before rendering something that depends on its size or position.

https://github.com/oliviertassinari/react-swipeable-views/issues/453#issue-348709038

To avoid the issue, we read the DOM in the render method once the first reconciliation is done. We can't do it during the first reconciliation because it can trigger a mismatch warning.

> I'm not familiar with SSR but do they really throw away the hole tree per request?

Yes, ssr starts from scratch for each request.

>  What if the server keeps the component tree mounted and just runs reconciliation for every request.

It's not what happens now. A problem for our future self, I doubt it will change. For ssr we compute the value during the first render path. 

> This is now only usable once per application because of the singleton. Any additional usage will potentially consider itself mounted when it's not.

It shouldn't be an issue, both variables will turn true after their first render.

> I have never seen the singleton components in react which is the only use case for this hook.

I have another example, something I would like to add in the core alongside `NoSrr`: `SsrOnly`:

```jsx
let reconciled = true

class SsrOnly extends React.Component {
  componentDidMount() {
    reconciled = false
  }

  render() {
    return reconciled ? this.props.children : true
  }
}

SsrOnly.propTypes = {
  children: PropTypes.node.isRequired,
}
```

Reply:
TL;DR: I have never seen this pattern and I'm not comfortable adding perf optimization with uncommon patterns (especially singletons), implementation details like react reconciliation and without benchmarks. If a unconventional implementation is chosen it should have thorough tests with actual use cases and not arbitrary unmounting. 

A rerender is not wasteful if it requires DOM reads. The current implementation now calls `window.matchMedia` on every render. Why is that not wasteful? 

>  We can't do it during the first reconciliation because it can trigger a mismatch warning.

During hydration? Well obviously `matchMedia` might not match between client and server. This should be handled gracefully. Could you provide an example that requires this singleton?

I see that my implementation doesn't call `matchMedia` on the server at all. So we would actually need to explicitly use a `ssrMatchMedia` and use that on the server during render.

Reply:
> I'm not comfortable adding perf optimization with uncommon patterns

I'm more than happy to try it out.

> Why is that not wasteful?

Let's say we don't add this logic.
`window.matchMedia` will be called on the did mount event, before the screen update, if the `matches` value is different, it will call setState, re-rendering the component. At this point, the screen wasn't updated yet, we have render twice and called `window.matchMedia`.
I think that we should **value** the number of renders way more than the call of `window.matchMedia`. `window.matchMedia` is CPU bounded, a rerender is not, we might have a lot of child elements.

> The current implementation now calls window.matchMedia on every render. Why is that not wasteful?

This is wasteful. I will do a second iteration. 

> During hydration? Well obviously matchMedia might not match between client and server. This should be handled gracefully. Could you provide an example that requires this singleton?
I see that my implementation doesn't call matchMedia on the server at all. So we would actually need to explicitly use a ssrMatchMedia and use that on the server during render.

Yes, matchMedia is not called on the server. No, the usage of `ssrMatchMedia` is optional.